paimon-core/src/main/java/org/apache/paimon/operation/KeyValueFileStoreScan.java

```diff
@@ -97,23 +103,50 @@ protected boolean filterByStats(ManifestEntry entry) {
 
     /** Note: Keep this thread-safe. */
     @Override
-    protected boolean filterWholeBucketByStats(List<ManifestEntry> entries) {
+    protected List<ManifestEntry> filterWholeBucketByStats(List<ManifestEntry> entries) {
+        if (valueFilter == null) {
+            return entries;
+        }
+
         // entries come from the same bucket, if any of it doesn't meet the request, we could filter
         // the bucket.
-        if (valueFilter != null) {
-            for (ManifestEntry entry : entries) {
-                if (valueFilter.test(
-                        entry.file().rowCount(),
-                        entry.file()
-                                .valueStats()
-                                .fields(
-                                        fieldValueStatsConverters.getOrCreate(
-                                                entry.file().schemaId()),
-                                        entry.file().rowCount()))) {
-                    return true;
-                }
+        if (canSkipWholeBucket(entries)) {
+            return Collections.emptyList();
+        }
+
+        // If there is no custom sequence field, we can filter the whole bucket by the max level.
+        if (!hasCustomSequenceField) {
```

Comment:
I think the bool should be more complicated.
1. minSequenceNumber in non-max levels should larger than maxSequenceNumber in max levels.
2. Merge Engine is deduplicate.

Reply:
You can add some ITCase to cover this.